diff --git a/dist/visitor-keys.js b/dist/visitor-keys.js
index 2746b8cc680466c7305d1c45ab4fc2413ee2fe2d..e1170ad62703a2d838daa617617eff17a303f387 100644
--- a/dist/visitor-keys.js
+++ b/dist/visitor-keys.js
@@ -92,13 +92,13 @@ const additionalKeys = {
     ClassDeclaration: SharedVisitorKeys.ClassDeclaration,
     ClassExpression: SharedVisitorKeys.ClassDeclaration,
     Decorator: ['expression'],
-    ExportAllDeclaration: ['exported', 'source', 'assertions'],
-    ExportNamedDeclaration: ['declaration', 'specifiers', 'source', 'assertions'],
+    ExportAllDeclaration: ['exported', 'source', 'attributes'],
+    ExportNamedDeclaration: ['declaration', 'specifiers', 'source', 'attributes'],
     FunctionDeclaration: SharedVisitorKeys.Function,
     FunctionExpression: SharedVisitorKeys.Function,
     Identifier: ['decorators', 'typeAnnotation'],
     ImportAttribute: ['key', 'value'],
-    ImportDeclaration: ['specifiers', 'source', 'assertions'],
+    ImportDeclaration: ['specifiers', 'source', 'attributes'],
     ImportExpression: ['source', 'options'],
     JSXClosingFragment: [],
     JSXOpeningElement: ['name', 'typeArguments', 'attributes'],
